From e6316ac717ffd2e14bb010bb6113ccf745ed08bb Mon Sep 17 00:00:00 2001
From: Aishwarya R <quic_aisr@quicinc.com>
Date: Tue, 19 Mar 2024 16:05:34 +0530
Subject: [PATCH] wifi: ath12k: Skip caldata download during FTM mode
with module param

Currently checksum fails when caldata is corrupted and
that would lead to firmware assert. Add module params which
allows driver to skip the caldata download to recover the
eeprom with faulty calibration data. It is useful for initial
calibration. Interfaces cannot be brought up when skip_caldata
enabled since the device should be in FTM mode.

insmod ath12k.ko ftm_mode=1 skip_caldata=1

Signed-off-by: Aishwarya R <quic_aisr@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/qmi.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/qmi.c b/drivers/net/wireless/ath/ath12k/qmi.c
index 2e3e80a..7678866 100644
--- a/drivers/net/wireless/ath/ath12k/qmi.c
+++ b/drivers/net/wireless/ath/ath12k/qmi.c
@@ -24,6 +24,10 @@
 #define PLATFORM_CAP_PCIE_GLOBAL_RESET	0x08
 #define ATH12K_QMI_MAX_CHUNK_SIZE	2097152
 
+static bool ath12k_skip_caldata;
+module_param_named(skip_caldata, ath12k_skip_caldata, bool, 0444);
+MODULE_PARM_DESC(ath12k_skip_caldata, "Skip caldata download");
+
 bool ath12k_cold_boot_cal = 1;
 module_param_named(cold_boot_cal, ath12k_cold_boot_cal, bool, 0644);
 MODULE_PARM_DESC(cold_boot_cal,
@@ -4918,7 +4922,15 @@ static int ath12k_qmi_load_bdf_qmi(struct ath12k_base *ab,
 		}
 		break;
 	case ATH12K_QMI_BDF_TYPE_CALIBRATION:
-
+		if (ath12k_skip_caldata) {
+			if (ath12k_ftm_mode) {
+				ath12k_warn(ab, "Skipping caldata download in FTM mode\n");
+				goto out;
+			}
+			ath12k_err(ab, "failed to skip caldata download. FTM mode is not enabled\n");
+			ret = -EOPNOTSUPP;
+			goto out;
+		}
 		if (ab->qmi.target.eeprom_caldata) {
 			file_type = ATH12K_QMI_FILE_TYPE_EEPROM;
 			tmp = filename;
-- 
2.34.1

